Adding support for provisioning of Windows VMs with enabled WinRM

ygy opened the pull request:
- Provisioning using pre-existing KeyVault
- Enabling WinRM with pre-existing certificate

andreaturli commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/OSProfile.java:
```diff
@@ -91,15 +90,58 @@ public static LinuxConfiguration create(final String disablePasswordAuthenticati
 
       @AutoValue
       public abstract static class WinRM {
+          public enum ListenerProtocol {
```
is `Protocol` a better name?

andreaturli commented on azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/VirtualMachineApiLiveTest.java:
```diff
@@ -96,6 +107,11 @@ public void setup() {
 
       vmName = String.format("%3.24s", System.getProperty("user.name") + RAND + this.getClass().getSimpleName()).toLowerCase().substring(0, 15);
+
+      vaultResourceGroup = System.getProperty("test.vault.resource.group");
```
I think this makes the test harder to execute: either you add it to `BaseAzureComputeApiLiveTest` or the test needs to create those prelimiary resources during the setup and destroy during tearDown. wdyt?

aledsage commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/OSProfile.java:
```diff
@@ -91,15 +90,58 @@ public static LinuxConfiguration create(final String disablePasswordAuthenticati
+              }
+
+              public static ListenerProtocol fromValue(String value) {
+                 ListenerProtocol[] items = ListenerProtocol.values();
```
Could use the same pattern as is in https://github.com/jclouds/jclouds-labs/blob/master/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/ComputeNode.java#L28

ygy commented:
@andreaturli thanks for the review. I've addressed the comments. The live test receives the parameters via system properties, because they have to exist before the test execution. Azurecompute-arm still can not provision keyvaults and secrets.

aledsage commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/OSProfile.java:
```diff
@@ -91,15 +90,58 @@ public static LinuxConfiguration create(final String disablePasswordAuthenticati
+              }
+
+              public static Protocol fromValue(String value) {
+                 Protocol[] items = Protocol.values();
```
Could use the same pattern as https://github.com/jclouds/jclouds-labs/blob/master/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/ComputeNode.java#L28, i.e.

    return (Protocol) GetEnumValue.fromValueOrDefault(value, Protocol.UNRECOGNIZED);

aledsage commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/VaultCertificate.java:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * Certificate's store name
+     */
+    public abstract String certificateStore();
```
Needs `@Nullable` - both here, and on the arg to `create` - see https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/virtualmachines-create-or-update#bk_vaultcert, which says that it is not required.

I believe that if you don't include `@Nullable` then `AutoValue`'s auto-generated code will throw a `NullPointerException` if the value is null.

(Actually, probably don't need it on the arg to `create()` - we don't do that elsewhere).

aledsage commented on azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/VirtualMachineApiMockTest.java:
A slightly more accurate comment would be:

    // See https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/virtualmachines-create-or-update
    // for where part of the example json response comes from. Unfortunately examples in the microsoft docs
    // are not valid json (e.g. missing commas, illegal quotes). Therefore this example merges the original 
    // real-world example (presumably taken from the jclouds wire log), and snippets from the microsoft docs.

bostko commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/OSProfile.java:
```diff
@@ -91,15 +90,52 @@ public static LinuxConfiguration create(final String disablePasswordAuthenticati
+
+              HTTP("http"),
+              HTTPS("https"),
+              UNRECOGNIZED("Unrecognized");
```
I cannot find any "Unrecognized" protocol in Azure documentation.

aledsage replied:
@bostko I think this is just a jclouds convention, of including "UNRECOGNIZED" for if we come across an unknown value (rather than it setting the field to null or throwing an exception).

bostko replied:
I do not think it is appropriate for this case since there are no other options.
Even if there are it will be invalid for any jclouds user.

aledsage replied:
@bostko I think it's better to produce an UNRECOGNIZED value rather than an exception or null, if we get back some different value. For example, if microsoft add a third possible value in their API, which jclouds doesn't understand, then we'll say that is `UNRECOGNIZED`.

bostko commented on azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/internal/BaseAzureComputeApiLiveTest.java:
```diff
@@ -82,6 +87,20 @@ protected void tearDown() {
+      // They have to be externally provided, because azurecompute-arm doesn't support creating KeyVaults yet
+      //
+      // TODO Replace the used configurations once full KeyVault implementation is added to azurecompute-arm
+      vaultResourceGroup = System.getProperty("test.vault.resource.group");
```
Rename to `test.azurecompute-arm.vault.resource.group`.

neykov commented:
Thanks @ygy and all reviewers. Changes look good. Merging.

neykov commented:
Merged to [master](https://git1-us-west.apache.org/repos/asf?p=jclouds-labs.git;a=commit;h=6e0cbc915cfb1e3f0cb9cdafd3b56b948e0e80a8).

aledsage commented:
Thanks @ygy (and @neykov @andreaturli @bostko)!

Could one of you please add this to the 2.0.x branch as well?

neykov commented:
Added to [2.0.x](https://git1-us-west.apache.org/repos/asf?p=jclouds-labs.git;a=commit;h=c4a91d2d909a4edf52adf1367b5353d81cde681b).
